nav2_collision_monitor/src/polygon.cpp

```diff
@@ -203,15 +225,9 @@ void Polygon::publish() const
     throw std::runtime_error{"Failed to lock node"};
   }
 
-  // Fill PolygonStamped struct
-  std::unique_ptr<geometry_msgs::msg::PolygonStamped> poly_s =
-    std::make_unique<geometry_msgs::msg::PolygonStamped>();
-  poly_s->header.stamp = node->now();
-  poly_s->header.frame_id = base_frame_id_;
-  poly_s->polygon = polygon_;
-
-  // Publish polygon
-  polygon_pub_->publish(std::move(poly_s));
+  // Actualize the time to current and publish the polygon
+  polygon_.header.stamp = node->now();
+  polygon_pub_->publish(polygon_);
```

Comment:
Yes, de-facto standard of publishing in Nav2 - it is a transferring resources to published message to another object.
However, the `polygon_` structure in our case is being used for two purposes: 1. storing polygon vertices and polygon frame data for publishing and 2. messages publishing. The storing some data into polygon (1) do much less frequently than publishing (2) polygon with updated time.
Since we need some structure for polygon points and frame storing (be it `PolygonStamped polygon_` or something else), we can not transfer the ownership to another object to it, since the stored object will be re-used in next cycles. Therefore, the way to `publish(std::move(unique_object))` - is to make a new `unique_object` and copy all the data into it. Something like:
```
auto unique_object = std::make_unique<PolygonStamped>(polygon_);
polygon_pub_->publish(std::move(unique_object));
```
However, what is the difference between the construction from above and just publishing the `polygon_` directly?
```
polygon_pub_->publish(polygon_);
```

During the publishing without std::move, ROS2 will allocate the memory for new object and make the copy of `polygon_` into it. The same as in the first example. If so, is there any performance reason to make these changes?

Reply:
I forget the exact rationale, but when publishing pointclouds in that fashion in developing the Ouster Drivers, Tom from Box Robotics pointed out that by giving it a smart pointer and moving its ownership to the publisher, the jitter the publication rate possible was significantly was reduced. But the idea is we're creating an object containing the data as a smart pointer and then handing it over without storing it for future iterations. I want to also say that you have to give it a smart pointer for IPC optimizations to work properly - which we don't use right now due to restrictions on QoS settings possible, but may in the future.

Reply:
Okay, added new object and ownership movement from it in the latest commit:
https://github.com/ros-planning/navigation2/pull/3245/files#diff-e7fd1328f82b3191cac8b30062c6eee0ea702a0be381dde8572120fa21a87595R237-R238